### PR TITLE
Memoryplanner

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: experimental/xformer/Analysis/MemoryPlan.cpp [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+  Block *block = &op->getRegion(0).front();
+
+  const LivenessBlockInfo *lvb = liveness.getLiveness(block);
+  for (auto v : values) {
+    Operation *startOp = lvb->getStartOperation(v);
+    valueInfo[v].firstUsed = operationIds[startOp];
+    valueInfo[v].lastUsed = operationIds[lvb->getEndOperation(v, startOp)];
+  }
+
+  printf("\n\n");
```
Comment (andrewstanfordjason) [inline experimental/xformer/Analysis/MemoryPlan.cpp:91]: is this a mistake?

**Review by andrewstanfordjason — CHANGES_REQUESTED** [review]
I think you left the debug printing in by mistake

File: experimental/xformer/Analysis/MemoryPlan.cpp (hunk 1 above)
Comment (ghost) [reply experimental/xformer/Analysis/MemoryPlan.cpp:91]: I had kept those in there while working on the buffer overlapping changes. Figured I'll get rid of them once that is also working. Do you think I should remove them now?

Comment (andrewstanfordjason) [reply experimental/xformer/Analysis/MemoryPlan.cpp:91]: no, that its a good reason

File: experimental/xformer/Analysis/MemoryPlan.cpp [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+
+  for (auto v : values) {
+    printf("\nvalue %d size = %d start op = %d end op = %d", valueInfo[v].id,
+           valueInfo[v].size, valueInfo[v].firstUsed, valueInfo[v].lastUsed);
+  }
+  printf("\n\n");
+}
+
+int MemoryPlan::getMaxMemoryUsed() {
+  Block *block = &op->getRegion(0).front();
```
Comment (andrewstanfordjason) [inline experimental/xformer/Analysis/MemoryPlan.cpp:101]: do you need to check the region count >0?

File: experimental/xformer/Analysis/MemoryPlan.cpp [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+  // Insert values and their sizes into priority queue
+  for (auto v : values) {
+    if (!outInVals.count(v) && !valueInfo[v].isConstant) {
+      queue.push({v, valueInfo[v].size});
+    }
+  }
+
+  ValuesOrderedByOffset allocatedValues;
+  auto v = queue.top().first;
+  queue.pop();
```
Comment (andrewstanfordjason) [inline experimental/xformer/Analysis/MemoryPlan.cpp:228]: pop_front() is a nice way to do this

**Review by andrewstanfordjason — APPROVED** [review]

File: experimental/xformer/Analysis/MemoryPlan.cpp (hunk 3 above)
Comment (ghost) [reply experimental/xformer/Analysis/MemoryPlan.cpp:228]: Didn't knnow about that. Will change.

File: experimental/xformer/Analysis/MemoryPlan.cpp (hunk 2 above)
Comment (ghost) [reply experimental/xformer/Analysis/MemoryPlan.cpp:101]: I don't think we need to, but will add that anyway. Would be cleaner.

**Review by scott-xmos — APPROVED** [review]

File: experimental/xformer/Analysis/MemoryPlan.cpp (hunk 2 above)
Comment (ghost) [reply experimental/xformer/Analysis/MemoryPlan.cpp:101]: Actually, we already do an assert for the region count in the constructor.

**Pull request merged**